### PR TITLE
Implement anydata type descriptor

### LINKED ISSUE #714: Implement `anydata` type descriptor
This is trivial after we have done #571 and #713.

It requires supporting `anydata` as a keyword.

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Implement `anydata` type descriptor
Resolves #714.

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: docs/subset12.md [hunk 1]
```diff
@@ -15,6 +15,7 @@
    * predefined basic type name: `boolean`, `decimal`, `error`, `float`, `int`, `string`
    * nil type `()`
    * `any` type
+   * `anydata` type
```
Comment (jclark) [inline docs/subset12.md:18]: Also mention in "Additions" section

File: compiler/modules/bir/verify.bal [hunk 2]
```diff
@@ -15,7 +15,7 @@ class VerifyContext {
         t:Context tc  = mod.getTypeContext();
         self.tc = tc;
         self.defn = defn;
-        self.anydataType = t:createAnydata(tc.env);
+        self.anydataType = t:createAnydata(tc);
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: We can get rid of `anydataType` field now that we are memoizing anydata in the type context.

File: compiler/modules/types/core.bal [hunk 3]
```diff
@@ -1210,12 +1211,18 @@ public function createJson(Env env) returns SemType {
...
+    SemType? memo = context.anydataMemo;
+    if memo != () {
+        return memo;
+    }
     ListDefinition listDef = new;
     MappingDefinition mapDef = new;
-    SemType ad = union(union(SIMPLE_OR_STRING, union(XML, TABLE)), union(listDef.getSemType(env), mapDef.getSemType(env)));
-    _ = listDef.define(env, rest = ad);
-    _ = mapDef.define(env, [], ad);
+    SemType tableTy = tableContaining(mapDef.getSemType(context.env));
```
Comment (jclark) [inline compiler/modules/types/core.bal]: Create an `env` local variable.

File: compiler/modules/nback/init.bal [hunk 4]
```diff
@@ -331,8 +331,8 @@ function createSubtypeStruct(InitModuleContext cx, t:UniformTypeCode typeCode, t
     if typeCode == t:UT_LIST_RW {
         return createListSubtypeStruct(cx, semType); 
     }
-    else if typeCode == t:UT_MAPPING_RW {
-        return createMappingSubtypeStruct(cx, semType); 
+    else if typeCode == t:UT_MAPPING_RW || typeCode == t:UT_TABLE_RO || typeCode == t:UT_TABLE_RW {
```
Comment (jclark) [inline compiler/modules/nback/init.bal]: Huh?

**Review by jclark — CHANGES_REQUESTED** [review]

**Comment by jclark** [comment]
The right fix to init.bal is to deal with the table uniform types in addComplexTypeDefn in the same way as the RO types

File: docs/subset12.md (hunk 1 above)
Comment (heshanpadmasiri) [reply docs/subset12.md:18]: fixed

File: compiler/modules/bir/verify.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/bir/verify.bal]: fixed

File: compiler/modules/types/core.bal (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/modules/types/core.bal]: fixed

File: compiler/modules/nback/init.bal (hunk 4 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/init.bal]: fixed

**Review by jclark — CHANGES_REQUESTED** [review]
Still need to fix resolveBuiltinTypeDesc

**Comment by heshanpadmasiri** [comment]
> Still need to fix resolveBuiltinTypeDesc

fixed

**Pull request merged**